rapids-cmake/cpm/nvtx3.cmake

```diff
@@ -0,0 +1,107 @@
+
+Allow projects to find `NVTX3` via `CPM` with built-in tracking of dependencies
+for correct export support.
+
+Uses the version of nvtx3 :ref:`specified in the version file <cpm_versions>` for consistency
+across all RAPIDS projects.
+
+.. code-block:: cmake
+
+  rapids_cpm_nvtx3( [BUILD_EXPORT_SET <export-name>]
+                    [INSTALL_EXPORT_SET <export-name>]
+                    [<CPM_ARGS> ...])
+
+.. |PKG_NAME| replace:: nvtx3
+.. include:: common_package_args.txt
+
+Result Targets
+^^^^^^^^^^^^^^
+  nvtx3-c, nvtx3-cpp targets will be created
+
+Result Variables
+^^^^^^^^^^^^^^^^
+  :cmake:variable:`NVTX3_SOURCE_DIR` is set to the path to the source directory of nvtx3.
+  :cmake:variable:`NVTX3_BINARY_DIR` is set to the path to the build directory of nvtx3.
+  :cmake:variable:`NVTX3_ADDED`      is set to a true value if nvtx3 has not been added before.
+  :cmake:variable:`NVTX3_VERSION`    is set to the version of nvtx3 specified by the versions.json.
+
+#]=======================================================================]
+function(rapids_cpm_nvtx3)
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.cpm.nvtx3")
+
+  set(options)
+  set(one_value USE_PROPRIETARY_BINARY BUILD_EXPORT_SET INSTALL_EXPORT_SET)
+  set(multi_value)
+  cmake_parse_arguments(_RAPIDS "${options}" "${one_value}" "${multi_value}" ${ARGN})
+
+  # Fix up _RAPIDS_UNPARSED_ARGUMENTS to have EXPORT_SETS as this is need for rapids_cpm_find
+  if(_RAPIDS_INSTALL_EXPORT_SET)
+    list(APPEND _RAPIDS_EXPORT_ARGUMENTS INSTALL_EXPORT_SET ${_RAPIDS_INSTALL_EXPORT_SET})
+  endif()
+  if(_RAPIDS_BUILD_EXPORT_SET)
+    list(APPEND _RAPIDS_EXPORT_ARGUMENTS BUILD_EXPORT_SET ${_RAPIDS_BUILD_EXPORT_SET})
+  endif()
+  set(_RAPIDS_UNPARSED_ARGUMENTS ${_RAPIDS_EXPORT_ARGUMENTS})
+
+  set(to_install OFF)
+  if(INSTALL_EXPORT_SET IN_LIST ARGN)
+    set(to_install ON)
+  endif()
+
+  include("${rapids-cmake-dir}/cpm/detail/package_details.cmake")
+  rapids_cpm_package_details(nvtx3 version repository tag shallow exclude)
+
+  include("${rapids-cmake-dir}/cpm/detail/generate_patch_command.cmake")
+  rapids_cpm_generate_patch_command(nvtx3 ${version} patch_command)
+
+  include("${rapids-cmake-dir}/cpm/find.cmake")
+  rapids_cpm_find(NVTX3 ${version} ${ARGN}
+                  GLOBAL_TARGETS nvtx3-c nvtx3-cpp
+                  CPM_ARGS
+                  GIT_REPOSITORY ${repository}
+                  GIT_TAG ${tag}
+                  GIT_SHALLOW ${shallow} ${patch_command} SOURCE_SUBDIR c
+                  EXCLUDE_FROM_ALL ${exclude})
+
+  include("${rapids-cmake-dir}/cpm/detail/display_patch_status.cmake")
+  rapids_cpm_display_patch_status(NVTX3)
+
+  # Set up install rules
+  if(to_install)
+    include(GNUInstallDirs)
+    install(DIRECTORY "${NVTX3_SOURCE_DIR}/c/include/" DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+    install(TARGETS nvtx3-c nvtx3-cpp EXPORT ${_RAPIDS_BUILD_EXPORT_SET})
+    if(NOT _RAPIDS_BUILD_EXPORT_SET STREQUAL _RAPIDS_INSTALL_EXPORT_SET)
```

Comment:
If a user has requested a `BUILD_EXPORT_SET` we need to generate the `export()` rules and not the install rules.

Reply:
@robertmaynard I might need more hand-holding here. I don't know what this is asking for.

Reply:
Basically this:

```cmake
  # Set up install rules
  if(_RAPIDS_BUILD_EXPORT_SET)
    install(TARGETS nvtx3-c nvtx3-cpp EXPORT ${_RAPIDS_BUILD_EXPORT_SET})
  endif()
  if(_RAPIDS_INSTALL_EXPORT_SET)
    include(GNUInstallDirs)
    install(DIRECTORY "${NVTX3_SOURCE_DIR}/c/include/" DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
    install(TARGETS nvtx3-c nvtx3-cpp EXPORT ${_RAPIDS_INSTALL_EXPORT_SET})
  endif()
```

Reply:
Need to also disable install rules when `EXCLUDE_FROM_ALL` flag has been set. You would look to have something like:

```cmake
  # Set up install rules
  install(TARGETS nvtx3-c nvtx3-cpp EXPORT nvtx3-targets)
  if(_RAPIDS_BUILD_EXPORT_SET)
    rapids_export(BUILD nvtx3
      EXPORT_SET nvtx3-targets
      GLOBAL_TARGETS  nvtx3-c nvtx3-cpp
      NAMESPACE nvtx::
      )
  endif()
  if(_RAPIDS_INSTALL_EXPORT_SET AND NOT exclude)
    include(GNUInstallDirs)
    install(DIRECTORY "${NVTX3_SOURCE_DIR}/c/include/" DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
    rapids_export(INSTALL nvtx3
      EXPORT_SET nvtx3-targets
      GLOBAL_TARGETS  nvtx3-c nvtx3-cpp
      NAMESPACE nvtx::
      )
  endif()
```

Reply:
Ah - so that's what the `exclude` was for. Thanks, we weren't sure.